File: scripts/rate-proxy.js
const http = require("http");
const https = require("https");

const PORT = Number(process.env.RATE_PORT) || 8787;
// Usa el mismo valor del gráfico: https://www.tradingview.com/chart/o5jQyVQx/?symbol=FX_IDC%3AUSDEUR
const TRADINGVIEW_URL = "https://scanner.tradingview.com/forex/scan";
const CACHE_TTL_MS = 60 * 1000;

let cache = {
  rate: null,
  fetchedAt: 0
};

const corsHeaders = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "GET, OPTIONS",
  "Access-Control-Allow-Headers": "Content-Type"
};

const fetchTradingView = (symbol) =>
  new Promise((resolve, reject) => {
    const payload = JSON.stringify({
      symbols: { tickers: [symbol], query: { types: [] } },
      columns: ["close"]
    });
    const req = https.request(
      TRADINGVIEW_URL,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Content-Length": Buffer.byteLength(payload),
          "User-Agent":
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120 Safari/537.36"
        }
      },
      (res) => {
        let data = "";
        res.on("data", (chunk) => {
          data += chunk;
        });
        res.on("end", () => {
          resolve(data);
        });
      }
    );
    req.on("error", reject);
    req.write(payload);
    req.end();
  });

const parseTradingViewRate = (jsonText) => {
  try {
    const data = JSON.parse(jsonText);
    const rate = data?.data?.[0]?.d?.[0];
    return Number(rate);
  } catch (e) {
    return null;
  }
};

const getRate = async (pair) => {
  const now = Date.now();
  if (cache.rate && cache.pair === pair && now - cache.fetchedAt < CACHE_TTL_MS) {
    return cache.rate;
  }
  const symbol = pair === "USDEUR" ? "FX_IDC:USDEUR" : "FX_IDC:USDCOP";
  const raw = await fetchTradingView(symbol);
  const rate = parseTradingViewRate(raw);
  if (!Number.isFinite(rate) || rate <= 0) {
    throw new Error("No se pudo leer la tasa");
  }
  cache = { rate, fetchedAt: now, pair };
  return rate;
};

const server = http.createServer(async (req, res) => {
  if (req.method === "OPTIONS") {
    res.writeHead(204, corsHeaders);
    res.end();
    return;
  }

  if (req.url.startsWith("/api/rate")) {
    try {
      const url = new URL(req.url, `http://${req.headers.host}`);
      const pair = (url.searchParams.get("pair") || "USDCOP").toUpperCase();
      const rate = await getRate(pair);
      res.writeHead(200, { ...corsHeaders, "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          rate,
          source: "tradingview",
          pair,
          fetchedAt: new Date().toISOString()
        })
      );
    } catch (error) {
      res.writeHead(502, { ...corsHeaders, "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          error: String(error?.message || "Error"),
          source: "tradingview",
          fetchedAt: new Date().toISOString()
        })
      );
    }
    return;
  }

  res.writeHead(404, corsHeaders);
  res.end("Not Found");
});

server.listen(PORT, () => {
  console.log(`Rate proxy activo en http://localhost:${PORT}/api/rate`);
});

File: scripts/rate-proxy.py
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import json
import threading
import time
import urllib.request
import urllib.parse

PORT = 8787
CACHE_TTL_SECONDS = 30
TRADINGVIEW_URL = "https://scanner.tradingview.com/forex/scan"
REQUEST_TIMEOUT_SECONDS = 4
FETCH_TIMEOUT_SECONDS = REQUEST_TIMEOUT_SECONDS + 1

cache_rate = None
cache_time = 0
cache_pair = None


def fetch_rate_tradingview(symbol):
  payload = {
    "symbols": {"tickers": [symbol], "query": {"types": []}},
    "columns": ["close"],
  }
  data = json.dumps(payload).encode("utf-8")
  req = urllib.request.Request(
    TRADINGVIEW_URL,
    data=data,
    headers={
      "User-Agent": (
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120 Safari/537.36"
      ),
      "Content-Type": "application/json",
    },
  )
  with urllib.request.urlopen(req, timeout=REQUEST_TIMEOUT_SECONDS) as resp:
    payload = json.loads(resp.read().decode("utf-8", "ignore"))
    data_row = (payload.get("data") or [None])[0]
    if not data_row:
      return None
    values = data_row.get("d") or []
    if not values:
      return None
    return float(values[0])


def fetch_rate_with_timeout(symbol):
  result = {"rate": None, "error": None}

  def _run():
    try:
      result["rate"] = fetch_rate_tradingview(symbol)
    except Exception as exc:
      result["error"] = exc

  thread = threading.Thread(target=_run, daemon=True)
  thread.start()
  thread.join(FETCH_TIMEOUT_SECONDS)
  if thread.is_alive():
    return None, TimeoutError("Tiempo de espera agotado")
  if result["error"] is not None:
    return None, result["error"]
  return result["rate"], None


def get_rate(pair):
  global cache_rate, cache_time, cache_pair
  now = time.time()
  if cache_rate and cache_pair == pair and (now - cache_time) < CACHE_TTL_SECONDS:
    return cache_rate
  symbol = "FX_IDC:USDEUR" if pair == "USDEUR" else "FX_IDC:USDCOP"
  rate, error = fetch_rate_with_timeout(symbol)
  if error:
    raise error
  if rate is None:
    raise ValueError("No se pudo leer la tasa")
  cache_rate = rate
  cache_time = now
  cache_pair = pair
  return rate


class RateHandler(BaseHTTPRequestHandler):
  def _send_headers(self, status, content_type="application/json"):
    self.send_response(status)
    self.send_header("Content-Type", content_type)
    self.send_header("Access-Control-Allow-Origin", "*")
    self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")
    self.send_header("Access-Control-Allow-Headers", "Content-Type")
    self.end_headers()

  def do_OPTIONS(self):
    self._send_headers(204, "text/plain")

  def do_GET(self):
    path, _, query = self.path.partition("?")
    if path != "/api/rate":
      self._send_headers(404, "text/plain")
      self.wfile.write(b"Not Found")
      return
    try:
      params = urllib.parse.parse_qs(query)
      pair = (params.get("pair", ["USDCOP"])[0] or "USDCOP").upper()
      rate = get_rate(pair)
      self._send_headers(200)
      payload = {
        "rate": rate,
        "source": "tradingview",
        "pair": pair,
        "fetchedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
      }
      self.wfile.write(json.dumps(payload).encode("utf-8"))
    except Exception as exc:
      self._send_headers(502)
      payload = {
        "error": str(exc),
        "source": "tradingview",
        "fetchedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
      }
      self.wfile.write(json.dumps(payload).encode("utf-8"))


if __name__ == "__main__":
  server = ThreadingHTTPServer(("localhost", PORT), RateHandler)
  print(f"Rate proxy activo en http://localhost:{PORT}/api/rate")
  server.serve_forever()
